### PR TITLE
#1207: Add a basic Input component (closes #1207)

### LINKED ISSUE #1207: Add a basic Input component
[Project card](https://github.com/buildo/react-components/projects/1?fullscreen=true#card-8547126)

## requirements
{describe the new feature}

## specs
{optional: describe technical specs to implement this feature, if not obvious}

## misc
{optional: other useful info}

### PR DESCRIPTION
**Pull request by giogonzo**
Closes #1207

### tests performed

screen shot 2018-03-29 at 13 02 25
screen shot 2018-03-29 at 13 02 04

> A Test Plan is used to show what you tested to make sure your code works fine. You should write here all that you did to test, and provide some results of your testing.

> These results can be screenshots, query results, or even just pointers to unit tests that successfully passed.

- [x] Google Chrome
- [ ] Internet Explorer

> An example of partial test is a field displaying 4 possible values. If 3 values are tested, with screenshots, and 1 is not, then it should be mentioned here.

### CODE REVIEW

File: src/Input/Input.tsx [hunk 1]
```diff
@@ -0,0 +1,69 @@
...
+import { props, t, ReactChildren } from '../utils';
+import omit = require('lodash/omit');
+import InputChildren from 'react-input-children';
+import { ObjectOverwrite } from 'typelevel-ts';
+
+export type InputRequiredProps = ObjectOverwrite<React.HTMLProps<HTMLInputElement>, {
+  /** value */
+  value: string,
+  /** onChange */
+  onChange: (value: string) => void,
```
Comment (FrancescoCioria) [inline src/Input/Input.tsx]: is there a case where we want a static input with only `defaultValue`? (maybe in addition with `disabled` `true`)

if yes, maybe we should make `value` and `onChange` optional? what do you think?

File: src/Input/Input.tsx [hunk 2]
```diff
@@ -0,0 +1,69 @@
...
+    const {
+      _onChange: onChange
+    } = this;
+
+    const inputProps = {
+      ...omit(props, 'onChange'),
+      value,
+      onChange,
+      wrapper: {
+        className: cx('input', { 'empty-value': value === '' }, className)
```
Comment (FrancescoCioria) [inline src/Input/Input.tsx]: in other components (like `Dropdown`) we did the opposite: `has-value`

File: src/Input/input.scss [hunk 3]
```diff
@@ -0,0 +1,50 @@
...
+$height: 28px !default;
+$padding-h: 12px !default;
+$font-size: 14px !default;
+$font-weight: $brc-regular !default;
+$placeholder-color: $brc-coolGrey !default;
+$color: $brc-darkGrey !default;
+$border-color: $brc-silver !default;
+$color-focus: $brc-darkGrey !default;
+$border-color-focus: $brc-waterBlue !default;
+$disabled-opacity: .6 !default;
```
Comment (FrancescoCioria) [inline src/Input/input.scss]: this should be named `opacity-disabled` :P

File: src/Input/Input.tsx (hunk 1 above)
Comment (giogonzo) [reply src/Input/Input.tsx]: static input -> `value='string' & onChange=() => {}`
static input with "placeholder" style... 🤷‍♂️ 

I'd leave it as is for the moment

File: src/Input/Input.tsx (hunk 2 above)
Comment (giogonzo) [reply src/Input/Input.tsx]: done in last commit

File: src/Input/input.scss (hunk 3 above)
Comment (giogonzo) [reply src/Input/input.scss]: fixed!

File: src/Input/Input.tsx [hunk 4]
```diff
@@ -48,18 +56,34 @@ export class Input extends React.PureComponent<Input.Props> {
...
       value,
+      status,
+      wrapper = {},
       ...props
     } = this.props as InputDefaultedProps;
     const {
       _onChange: onChange
     } = this;
 
+    const isSuccess = status === InputStatusT('success');
```
Comment (FrancescoCioria) [inline src/Input/Input.tsx]: why `InputStatusT('...')`? what happened to TS? :P

Comment (giogonzo) [reply src/Input/Input.tsx]: 😅 copy-pasta

**Pull request merged**